design/login.md

```diff
@@ -0,0 +1,103 @@
+    "description": "Default DC/OS authenticator"
+  }
+}
+```
+
+The above definition is a default DC/OS login provider, its ID is **dcos-users**.
+The ID has no special semantic apart from uniquely identifying a login provider.
+
+When the `/acs/api/v1/auth/providers` endpoint is not present, the available login provider is
+detected by making an unauthenticated `HEAD` request to a well-known protected DC/OS resource and
+reading the `WWW-Authenticate` header DC/OS returns in the 401 response:
+
+- `acsjwt` : the cluster only supports the `dcos-users` provider as defined above.
+- `oauthjwt` : the cluster only supports the OpenID implicit flow provider, as defined below:
+
+``` json
+{
+  "dcos-oidc-auth0": {
+    "authentication-type": "oidc-implicit-flow",
+    "client-method": "browser-prompt-authtoken",
+    "config": {
+      "start_flow_url": "/login?redirect_uri=urn:ietf:wg:oauth:2.0:oob"
+    },
+    "description": "Google, GitHub, or Microsoft"
+  }
+}
+```
+
+## Log in to a DC/OS cluster
+
+Log in to a DC/OS cluster is the action of presenting some credentials to the cluster in order to
+get an authorization token in return (also known as `ACS token`).
+
+The DC/OS CLI supports 6 login providers types (`authentication-type`):
+
+- **dcos-uid-password** : Log in using a standard DC/OS user account (username and password).
+- **dcos-uid-password-ldap** : Log in using an LDAP user account (username and password).
+- **dcos-uid-servicekey** : Log in using a DC/OS service user account (username and private key).
+- **saml-sp-initiated** : Log in using SAML 2.0.
+- **oidc-authorization-code-flow** : Log in using OpenID Connect authorization code flow.
+- **oidc-implicit-flow** : Log in using OpenID Connect implicit flow.
+
+A login happens through the `dcos auth login` command, which accepts the following flags:
+
+- **--provider** (string) : Specify the login provider ID to use.
+- **--username** (string) : Specify the username.
+- **--password** (string) : Specify password in plaintext (insecure).
+- **--password-file** (string) : Specify the path to a file that contains the password (insecure).
+- **--password-env** (string) : Specify an environment variable name that contains the password.
+- **--private-key** (string) : Specify the path to a file that contains the private key for service
+    account login.
+
+(A login also happens at the end of the `dcos cluster setup` command, which accepts all the flags
+from the `dcos auth login` command.)
+
+In order to perform a login, the CLI must select a login provider to use. This selection can occur
+in 3 different ways:
+
+- **explicitly** : the user passed the `--provider` flag with a given login provider ID.
+- **implicitly** : a single login provider is available or the user passed some flags
+    which are specific to a single login provider.
+- **manually** : when the login provider is not explicit nor implicit, a list is prompted to let
+    the user select a login provider manually.
```

Comment:
Is there a flag to prevent interactivity (e.g. the manual option here) for scripts?

Reply:
Do you mean a generic flag or one for this specific case? What I would do here in scripts is passing the --provider flag explicitly.

Reply:
A command stopping for user interaction can sometimes be hard to debug, especially when it occurs in a batch process on a remote node (e.g. CI). It is nicer to have an option to ensure that if the script gets to a point where it needs interaction, it will fail with a non-zero exit code and an informative message instead.

So I think it would be generic. If the user forgets to set where the password is coming from, or what provider to use, then the script fails rather than hanging waiting for a non-existent interactive user.

However, this is particularly important for a case like this where a script may work for me (because I only have access to a single login provider) but fail for you (because you have access to multiple login providers).